### PR TITLE
Finite counterexamples generated by the Vampire theorem prover

### PR DESCRIPTION
**Pull request by Command-Master**
Solves ~8,000 unknown implications

### CODE REVIEW

**Comment by zaklogician** [comment]
Very nice!

Note: we had an accidental regression in `All4x4Tables` after #271 (my review wasn't careful enough, so a different, older set of refutations ended up being pushed). So if you ran this in the last 3 hours, some of the anti-implications might already have counterexamples in `All4x4Tables` which will be restored by #275.

**Comment by Command-Master** [comment]
> Very nice!
> 
> Note: we had an accidental regression in `All4x4Tables` after #271 (my review wasn't careful enough, so a different, older set of refutations ended up being pushed). So if you ran this in the last 3 hours, some of the anti-implications might already have counterexamples in `All4x4Tables` which will be restored by #275.

I ran it before that, so it should be fine

**Comment by zaklogician** [comment]
Btw: having taken a look, it looks like Vampire managed to generate a large number of different tables. Is each example constructed independently, or are there some checks to see whether the resulting tables refute any implications apart from the one in the problem statement?

If the former, you can create an issue for this after the merge: I'll eventually get around to recovering the generated tables and putting them into `All4x4Magmas` so that we'll know all the implications they refute.

**Comment by nomeata** [comment]
> Any idea what might be causing the `(kernel) deterministic timeout`? It didn't happen on my computer

Proving that a `Fin 10` magma satisfies some equation with a few variables may be too slow to do for the `decideFin!` tactic.

Maybe try `by native_decide` instead for large magmas and small number of facts (only 2 here).

This may need some more engineering to make tractable.

**Comment by Command-Master** [comment]
I've noticed that MemoFinOp is using a base which isn't a power of 2, I could try to change that (and make it use bitshifts and bitwise and) and see if that helps

**Comment by nomeata** [comment]
I have strong doubts that this is the problem. It's still two operations, so the same deterministic timeout would apply. In fact, with shift you need an extra operation to test/clamp to `Fin n`.

I think it’s just the sheer number of operations.

Profiling first would help.

File: equational_theories/MemoFinOp.lean [hunk 1]
```diff
@@ -9,39 +9,41 @@ namespace MemeFinOp
...
-def opOfTable {n : Nat} (t : Nat) (a : Fin n) (b : Fin n) : Fin n :=
-  let i := a.val * n + b.val
-  let r := (t / n^i) % n
-  ⟨r, Nat.mod_lt _ (Fin.pos a)⟩
+def opOfTable {n : Nat} (s t : Nat) (a : Fin n) (b : Fin n) : Fin n :=
+  ⟨(t >>> (s * (a.val * n + b.val))) % n, Nat.mod_lt _ (Fin.pos a)⟩
 
 /-- `enum n` is the array of all elements of `Fin n` in order -/
-private def Fin.enum (n) : Array (Fin n) := Array.ofFn id
+private def Fin.enum (n) : Array (Fin n) := Array.ofFn Fin.rev
```
Comment (nomeata) [inline equational_theories/MemoFinOp.lean:16]: You probably should update the docstring as well (or remove it for this private helper)

Comment (Command-Master) [reply equational_theories/MemoFinOp.lean]: I undid this change, indeed it didn't help

File: .gitattributes [hunk 2]
```diff
@@ -10,4 +10,4 @@ equational_theories/Generated/FinSearch/theorems/*.lean linguist-generated
 equational_theories/Generated/FinSearch.lean linguist-generated
 equational_theories/Generated/MagmaEgg/*/*.lean linguist-generated
 equational_theories/Generated/MagmaEgg/*.lean linguist-generated
-equational_theories/Generated/VampireProven/Proofs*.lean linguist-generated
+equational_theories/Generated/VampireProven/*.lean linguist-generated
```
Comment (nomeata) [inline .gitattributes:13]:
```suggestion
equational_theories/Generated/VampireProven.lean linguist-generated
equational_theories/Generated/VampireProven/*.lean linguist-generated
```

File: scripts/transitive_reduction_nonimplications.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+# Given as an input the JSON produced by `lake exe extract_implications --json`, outputs
+# a transitive reduction of the non-implications
+
+require File.join(__dir__, 'graph')
+
+graph = Graph.new
+data = JSON.parse(File.read(ARGV[0]))
+data["implications"].each { |s|
+  graph.add_edge([false, s["rhs"][8,10].to_i], [false, s["lhs"][8,10].to_i])
```
Comment (nomeata) [inline scripts/transitive_reduction_nonimplications.rb:11]: Is this `[8,10]` extracting the number from the definition name? But what if the number is longer than two characters? Or is this something else?

Comment (Command-Master) [reply scripts/transitive_reduction_nonimplications.rb:11]: It takes a substring of length 10 starting from the eighth character

Comment (nomeata) [reply scripts/transitive_reduction_nonimplications.rb:11]: Ah, it's length, not end. TIL :-)

**Comment by zaklogician** [comment]
I can confirm that the minimization has resolved the OOM issues I experienced earlier, and I am now able to build tihs on my laptop.

**Pull request merged**